src/modules/Dashboard/features/Applet/Activities/Activities.tsx

```diff
@@ -1,73 +1,72 @@
 import { Spinner } from 'shared/components';
 import { useAsync } from 'shared/hooks';
-import {
-  ActivitiesData,
-  ActivityGrid,
-  useActivityGrid,
-} from 'modules/Dashboard/components/ActivityGrid';
+import { ActivityGrid, useActivityGrid } from 'modules/Dashboard/components/ActivityGrid';
 import { FlowGrid } from 'modules/Dashboard/components/FlowGrid';
 import { Activity, ActivityFlow } from 'redux/modules';
 import { StyledFlexColumn } from 'shared/styles';
 
 import { ActivitiesSectionHeader } from './ActivitiesSectionHeader';
 import { ActivitiesToolbar } from './ActivitiesToolbar';
 
+const dataTestId = 'dashboard-applet-activities';
+
 export const Activities = () => {
   const { appletId } = useParams();
   const { t } = useTranslation('app');
-  const [flows, setFlows] = useState<ActivityFlow[]>([]);
-  const [flowActivities, setFlowActivities] = useState<Activity[]>([]);
-  const [activitiesData, setActivitiesData] = useState<ActivitiesData | null>(null);
-  const dataTestId = 'dashboard-applet-activities';
-  const { formatRow, TakeNowModal } = useActivityGrid(dataTestId, activitiesData);
-
-  const { execute: getActivities, isLoading: isLoadingActivities } = useAsync(
-    getAppletActivitiesApi,
-    (response) => {
-      const activitiesDetails = response?.data.result.activitiesDetails;
-
-      return setActivitiesData({ result: activitiesDetails, count: activitiesDetails.length });
-    },
+  const {
+    execute: getActivities,
+    isLoading: isLoadingActivities,
+    value: activitiesResponse,
+    previousValue: previousActivitiesResponse,
+  } = useAsync(getAppletActivitiesApi);
+  const {
+    execute: getFlows,
+    isLoading: isLoadingFlows,
+    value: flowResponse,
+    previousValue: previousFlowResponse,
+  } = useAsync(getAppletApi);
+  const flows: ActivityFlow[] =
+    (flowResponse ?? previousFlowResponse)?.data?.result?.activityFlows ?? [];
+  const flowActivities: Activity[] =
+    (flowResponse ?? previousFlowResponse)?.data?.result?.activities ?? [];
+  const activities: Activity[] = useMemo(
+    () => (activitiesResponse ?? previousActivitiesResponse)?.data?.result?.activitiesDetails ?? [],
+    [activitiesResponse, previousActivitiesResponse],
   );
-
-  const { execute: getFlows, isLoading: isLoadingFlows } = useAsync(getAppletApi, (response) => {
-    if (response?.data?.result) {
-      setFlows(response.data.result.activityFlows ?? []);
-      setFlowActivities(response.data.result.activities ?? []);
-    }
+  const { formatRow, TakeNowModal } = useActivityGrid(dataTestId, {
+    result: activities,
+    count: activities.length,
   });
+  const formattedActivities = useMemo(
+    () => activities.map((activity) => formatRow(activity)),
+    [activities, formatRow],
+  );
 
   const isLoading = isLoadingActivities || isLoadingFlows;
+  const showContent =
+    (isLoading && previousActivitiesResponse?.data?.result?.activitiesDetails?.length > 0) ||
+    !isLoading;
```

Comment:
I appreciate all the effort you went to to improve the loading state UX! Who know it would end up becoming this complicated. 😅

I'm wondering if this loading state pattern, as you've implemented in the same way for both Activities screens, might make sense to – eventually – apply more broadly in other consumers of `useAsync`? Not that I think we should update all the screens to invoke this same logic yet. More that I see an opportunity to encapsulate these bits of logic into `useAsync`, rather than repeat for each screen, which feels just a little unwieldy.

- Perhaps `response ?? prevResponse` could be returned from the hook as, say, `cachedResponse`?
- Somehow encapsulating the logic in `showContent` in a more generic way, also returnable from the hook as, say, `hasCachedData`? (Does it need to delve into previous response data structure as you've done here, or is the truthiness of `prevResponse` enough?)

Just some thoughts to improve comprehension and reusability, not sure if it's practical! I'm totally fine to keep it as you've done it here.

Reply:
Yeah — this `useAsync` seems like a great target for building more patterns around, because I do appreciate the way it returns state directly in a way that lets us avoid hook or callback-driven approaches (eg an `onSuccess` that sets the response data locally in the component). 

I was searching around the codebase and there doesn't seem to be much use of it in the way I am here, outside of using the `isLoading` flag.

I'll say one of the first changes I'd want to make is returning the data object directly, instead of the response, to reduce the amount of property drilling going on 😅.

> Somehow encapsulating the logic in showContent in a more generic way, also returnable from the hook as, say, hasCachedData? (Does it need to delve into previous response data structure as you've done here, or is the truthiness of prevResponse enough?)?

This was something I was thinking about already — the interface of this hook looks a lot like Apollo's [useLazyQuery](https://www.apollographql.com/docs/react/api/react/hooks/#uselazyquery), specifically. They return a `called` value which is analogous to your suggestion, though that gets set as soon as the initial request starts. TanStack's [useQuery](https://tanstack.com/query/v4/docs/framework/react/reference/useQuery) includes more nuanced fields akin to what we're trying to do here, like `isInitialLoading` distinct from `isLoading`, and `isFetched`, which is essentially the check we're trying to do here (show whatever content we have, so long as it's not the initial load).

I think I'm not so inclined to make additional changes yet, but I think this is a good spot to have some eyes on going forward 👀👀.